packages/next/src/build/webpack/loaders/next-app-loader/index.ts

```diff
@@ -368,23 +395,57 @@ async function createTreeCodeFromPath(
       const normalizedParallelKey = normalizeParallelKey(parallelKey)
       let subtreeCode
       // If it's root not found page, set not-found boundary as children page
-      if (isNotFoundRoute && normalizedParallelKey === 'children') {
-        const notFoundPath =
-          definedFilePaths.find(([type]) => type === 'not-found')?.[1] ??
-          defaultHTTPAccessFallbackPaths['not-found']
-
-        const varName = `notFound${nestedCollectedDeclarations.length}`
-        nestedCollectedDeclarations.push([varName, notFoundPath])
-        subtreeCode = `{
-          children: [${JSON.stringify(UNDERSCORE_NOT_FOUND_ROUTE)}, {
-            children: ['${PAGE_SEGMENT_KEY}', {}, {
-              page: [
-                ${varName},
-                ${JSON.stringify(notFoundPath)}
-              ]
-            }]
-          }, {}]
-        }`
+      if (isNotFoundRoute) {
+        if (normalizedParallelKey === 'children') {
+          const matchedGlobalNotFound = isGlobalNotFoundEnabled
+            ? definedFilePaths.find(
+                ([type]) => type === GLOBAL_NOT_FOUND_FILE_TYPE
+              )?.[1] ?? defaultGlobalNotFoundPath
+            : undefined
+
+          // If custom global-not-found.js is defined, use global-not-found.js
+          if (matchedGlobalNotFound) {
+            const varName = `notFound${nestedCollectedDeclarations.length}`
+            nestedCollectedDeclarations.push([varName, matchedGlobalNotFound])
+            subtreeCode = `{
+              children: [${JSON.stringify(UNDERSCORE_NOT_FOUND_ROUTE)}, {
+                children: ['${PAGE_SEGMENT_KEY}', {}, {
+                  page: [
+                    ${varName},
+                    ${JSON.stringify(matchedGlobalNotFound)}
+                  ]
+                }]
+              }, {}]
+            }`
+          } else {
+            // If custom not-found.js is found, use it and layout to compose the page,
+            // and fallback to built-in not-found component if doesn't exist.
+            const notFoundPath =
+              definedFilePaths.find(([type]) => type === 'not-found')?.[1] ??
+              defaultNotFoundPath
+            const varName = `notFound${nestedCollectedDeclarations.length}`
+            nestedCollectedDeclarations.push([varName, notFoundPath])
+            subtreeCode = `{
+              children: [${JSON.stringify(UNDERSCORE_NOT_FOUND_ROUTE)}, {
+                children: ['${PAGE_SEGMENT_KEY}', {}, {
+                  page: [
+                    ${varName},
+                    ${JSON.stringify(notFoundPath)}
+                  ]
+                }]
+              }, {}]
+            }`
+          }
+        }
+      }
+
+      // For 404 route
+      // if global-not-found is in definedFilePaths, remove root layout for /_not-found
+      // TODO: remove this once global-not-found is stable.
```

Comment:
Maybe I'm misunderstanding the comment but shouldn't `not-found` routes always have the root layout applied and then if they need to escape the root layout they leverage the `global-not-found`?

Possibly a test case asserting the behavior of root layouts with this flag and `not-found` being shown would be good.

Reply:
`/_not-found` (aka 404 route) doesn't need layout anymore since global-not-found is required to have `<html>` and `<body>` tags. We have the test cases in `not-present.test.ts` with hitting the `/_not-found` route.

Also add a case of calling `notFound()` there after this comment.